File: solutions/go/allergies/1/allergies.go
package allergies

var allergens = map[string]uint{
	"eggs":         1,
	"peanuts":      2,
	"shellfish":    4,
	"strawberries": 8,
	"tomatoes":     16,
	"chocolate":    32,
	"pollen":       64,
	"cats":         128,
}

func Allergies(allergies uint) []string {
	result := make([]string, 0)
	for allergen := range allergens {
		if AllergicTo(allergies, allergen) {
			result = append(result, allergen)
		}
	}
	return result
}

func AllergicTo(allergies uint, allergen string) bool {
	flag := allergens[allergen]
	return allergies&flag != 0
}

File: solutions/elixir/allergies/1/lib/allergies.ex
import Bitwise

defmodule Allergies do
  @allergens ~w[eggs peanuts shellfish strawberries tomatoes chocolate pollen cats]
  @doc """
  List the allergies for which the corresponding flag bit is true.
  """
  @spec list(non_neg_integer) :: [String.t()]
  def list(flags) do
    @allergens |> Stream.filter(&allergic_to?(flags, &1)) |> Enum.to_list()
  end

  @doc """
  Returns whether the corresponding flag bit in 'flags' is set for the item.
  """
  @spec allergic_to?(non_neg_integer, String.t()) :: boolean
  def allergic_to?(flags, item) do
    (flags &&& 1 <<< Enum.find_index(@allergens, &(&1 == item))) != 0
  end
end
